navapbc/platform-test
# Add ability to check database roles configuration

## Ticket

n/a

## Changes
* Add "check" option to role manager lambda function
* Add check-database-roles.sh script to call role manager with "check" option
* Add make infra-check-database-roles target
* Add cluster_id output to database module
* Add db app and migrator access to role manager lambda function
* Update set-up-database.md instructions to check database roles at the end of setup

Other changes
* Print role manager's logs when running create-or-update-database-roles.sh

## Context for reviewers

Another developer experience thing similar to https://github.com/navapbc/template-infra/pull/415, this change allows someone to check that the database is properly configured without having to spin up the service layer. This change updates the role manager lambda function to have a "check" mode that logs in to the database as the "migrator" and "app" users (to make sure that IAM auth works), and attempts to perform queries that on the database to check that they have proper privileges. Specifically, the migrator user checks the search path to make sure that new tables are created in the right schema rather than public, it creates a table, then the app user writes something to that table, then the migrator user drops the table. This should improve the first time developer setup experience, and should also help with maintenance e.g. if someone needs to refactor the database code, they can do so and have some confidence that their changes won't break things.

## Testing

1. Create new database cluster in a new workspace `lycheckdb`
   <img width="613" alt="image" src="https://github.com/navapbc/platform-test/assets/447859/709bbcd3-be77-495a-aaee-116763860557">

   <img width="381" alt="image" src="https://github.com/navapbc/platform-test/assets/447859/4c69430d-d411-400c-baf8-2d2c0d5566c6">


2. Run database check function (this should fail since we didn't run configure database roles function yet)

   <img width="1045" alt="image" src="https://github.com/navapbc/platform-test/assets/447859/b8652e3a-7aab-4d62-b8e4-1ecda3b97f3c">

3. Run configure database roles function
   <img width="1044" alt="image" src="https://github.com/navapbc/platform-test/assets/447859/841d260c-691f-4c12-8cb0-20f3a0faa55f">

4. Run database check function again

   first time timed out
   <img width="1045" alt="image" src="https://github.com/navapbc/platform-test/assets/447859/dbd3f869-3e82-4b18-86a5-4474b7ed4bdd">

   retry was successful
   <img width="1043" alt="image" src="https://github.com/navapbc/platform-test/assets/447859/7fb406ce-300b-4ab8-a70c-6d68221c9c03">

5. Clean up (destroy db and delete workspace)

TODO once we're done with this PR
